### PR TITLE
UP-17: Investigate and upgrade rwandahivflowsheet

### PR DESCRIPTION
**Pull request by rubailly**
- Replaced `#getBestName` by `#getName`
- Replaced `#getDiscontinueDate` by `#getEffectiveStopDate`
- Replaced `#getStartDate` by `#getEffectiveStartDate`
- Replaced `#isCurrent` by `#isActive`
- Updated `#getEncounters`

- Replaced `#getUnits` by `#getDoseUnits().getDisplayString()` and `#drugOrder.getFrequency` by `drugOrder.getConcept().getDisplayString() `

- Added reportingcompatibility dependency and replaced `PatientSetService` with `ReportingCompatibilityService`
- Added a new implmentation for  `#getDrugOrdersByPatient`
- Added a new implmentation for `#getDiscontinuedReason`

### CODE REVIEW

File: api/pom.xml
Comment (mseaton) [inline api/pom.xml:12]: We should aim to keep all omod/web dependencies in the omod project, and have no such dependencies in the api project.  So this should be moved to the omod pom.xml

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ImageMapping.java
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ImageMapping.java:188]: I have been using o.getValueCoded().getDisplayString(), which might be a better choice.  Need to look at how they differ.  The main thing is you don't want this to ever return null or an empty string, which I think the getDisplayString() guards against.

Comment (mseaton) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ImageMapping.java:188]: This comment applies to all of the other places where displaying a concept is done this way in this PR

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java]: By using "getEffectiveStopDate", this method already returns either the dateStopped or the autoExpireDate, whichever is appropriate.  So you can get rid of this second check against drugOrder.getAutoExpireDate and just check the getEffectiveStopDate

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/UIHelper.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/UIHelper.java:86]: This is the best you can do for now, but this might not actually do what it is meant to do following the migration to coded units and frequencies, since this is pretty hacky code written specifically for names of free-text data.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java]: Same comment here as above, you can do away with the autoexpire check since you are using effective stop date

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java:259]: No idea here.  I doubt this is going to work as intended, and probably needs to be independently ticketed for someone to look into/address as part of the order entry work overall.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java]: Ideally you'd look this up once outside of the loop and not continuously look this up from the service.  Caching likely means it doesn't matter, but could be inefficient.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:204]: You may need to do a sanity check here to ensure that "isActive" is comparable to "isCurrent" previously.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/PediHivFlowsheetFormData.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/PediHivFlowsheetFormData.java]: Same comment here re: auto expire date

**Review by mseaton — CHANGES_REQUESTED** [review]
Excellent work.  A few comments in the review none of which are super critical, but I'd want you to have the chance to review and potentially address before I merge.  Let me know if you are happy or want to issue a further commit to address any of those comments and we can merge in.

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java:31]: `startDate` was replaced with `dateActivated` https://wiki.openmrs.org/x/yAgJAg

File: pom.xml
Comment (dkayiwa) [inline pom.xml]: Is there any reason why you are using a snapshot instead of released version of the module?

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/utils/Utils.java
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/utils/Utils.java]: I would use `drugOrder` as the parameter name instead of `dor`.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java]: How about using something like this? https://pastebin.com/x9PWZ3Q8

File: api/pom.xml
Comment (rubailly) [reply api/pom.xml:12]: The issue is the `org.openmrs.module.web.extension.LinkExt` class which is used by `RwandaAdultAndPediHIVFlowsheetFormItem`, `RwandaAdultHIVFlowsheetFormItem` and  `RwandaPediHIVFlowsheetFormItem` classes from the API project. I get compilation errors in those classes when I remove the legacy-ui dependency from the API project.

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ImageMapping.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ImageMapping.java:188]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/4048a80bbb356519bdd8a10b0ec226a093d4de74)

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/9478de14df21fce4aebb0e1cacfad3baf8b0c061)

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:210]: Shouldn't we be getting the discontinuedDate the same way we do get the discontinuedReason?

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java:120]: Are we intentionally ignoring the discontinuedDate check?

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/a2a8256b394ac8c5384adc841c152682cd856677)

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java]: I couldn't find a replacement for [getDrugOrdersByPatient](https://resources.openmrs.org/doc-1.9/org/openmrs/api/OrderService.html). The method has been letterly removed from core, and I was forced to get all orders for a patient and loop through all of them checking each one to see if it is actually a drug order.
Do you know a better method we can use here?

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/PediHivFlowsheetFormData.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/PediHivFlowsheetFormData.java]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/1f69833ead11342f2a50e00f60357f6c48a42a76)

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/utils/Utils.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/utils/Utils.java]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/a6169460219bcfed598fc0a6f7555d9816327f3a)

File: pom.xml
Comment (rubailly) [reply pom.xml]: [Fixed](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/301ca92371f8e11a20b9ba38d8d43770645c14a8)
Seems 2.0.6 is the latest release.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java]: Thanks a lot! [Here](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/49ca7831c7a2659a22130e2b8953c95addc2270b) is the commit.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java]: [Here](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/49ca7831c7a2659a22130e2b8953c95addc2270b) is the commit after reading @dkayiwa feedback

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java:120]: discontinuedDate is no longer available, so I'm using #getEffectiveStopDate which from the javadoc seems to handle this efficiently

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:210]: Seems #getEffectiveStopDate should handle this.

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java:31]: With the concept of Urgency though, #getEffectiveStartDate seems to be the best option here. It returns #scheduledDate if Urgency is Scheduled or #dateActivated if Urgency is not Scheduled.

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java:31]: Did you get a chance to look at what Nicholas Ingosi asked at the bottom of the link that i shared above?

File: api/pom.xml
Comment (mseaton) [reply api/pom.xml:12]: @rubailly - yes, well all of those FormItem classes that extend LinkExt should also be in the omod project.  Just move these into a package in the omod project, and change config.xml (which is also in omod) to refer to these new packages, and things should work fine.

Comment (rubailly) [reply api/pom.xml:12]: [Done](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/2b97a7e318a90cefe799f949a6644364654aec0a)

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java
Comment (rubailly) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/TbEpisodeMapping.java:31]: I hadn't seen it -sorry.
Here is the [fix](https://github.com/PIH/openmrs-module-rwandahivflowsheet/pull/2/commits/213f261e1a05d4051dc87143028a75f462b0b26d).

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/AdultHivFlowsheetFormData.java:259]: I'll test it on a 1.9 instance and see how it behaves and once we merge this PR, I'll test on a 2.3 instance to compare and create a ticket if things break.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (rubailly) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:204]: Saw the page at https://docs.openmrs.org/doc-1.10/org/openmrs/Order.html#isActive-java.util.Date- and thought #isActive might be the right one

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (mseaton) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java:120]: This seems right to me.

File: api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java:130]: Do we know that dateActivated is right here, and not effectiveStartDate?  Can we discuss?

Comment (mseaton) [reply api/src/main/java/org/openmrs/module/rwandahivflowsheet/impl/pih/ProphylaxisMapping.java:130]: This seems to be used throughout this PR, and I think I saw that @dkayiwa recommended this based on a thread somewhere.  I'd be interested to learn more.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/FormDataModel.java:460]: This line seems like it is dangling, and probably was supposed to be removed from a previous commit.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (mseaton) [reply omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:210]: Agreed.

File: omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/rwandahivflowsheet/web/model/HivFlowsheetFormData.java:220]: Just need to confirm that this doesn't need to handle orders scheduled for the future.

**Review by mseaton — APPROVED** [review]
Approving, pending resolution of a few minor issues I pointed out in the code.
My biggest question is whether dateActivated is the right date to use in place of startDate in the code, or whether effectiveStartDate is better.  Can we have this conversation over Slack or elsewhere prior to merging this in, just to confirm our approach @rubailly  and @dkayiwa ?

**Review by mseaton — APPROVED** [review]

**Comment by rubailly** [comment]
Thanks a lot @mseaton annd @dkayiwa 
I'm merging this now.

**Pull request merged**
